File: lesson-08-aspnet-core-web-apis/samples/L08.Rating.Vb/PremiumCalculator.vb
Imports Microsoft.Extensions.Options

''' <summary>The MotorQuote rating rules — the track's canonical illustrative tariff, not a real one.</summary>
Public NotInheritable Class PremiumCalculator
    Private ReadOnly _rates As RatingOptions

    Public Sub New(options As IOptions(Of RatingOptions))
        _rates = options.Value
    End Sub

#Region "calculate"
    Public Function Calculate(input As RatingInput) As PremiumBreakdown
        ' Past the last rung of the claims ladder the tariff declines the quote instead of pricing it.
        If input.ClaimsLast5Years >= _rates.ClaimLoadings.Length Then
            Return PremiumBreakdown.Declined(
                $"{_rates.ClaimLoadings.Length}+ claims in 5 years")
        End If

        Dim basePremium = input.SumInsured * BaseRateFor(input.Coverage)

        Dim loading = _rates.ClaimLoadings(input.ClaimsLast5Years)
        If input.DriverAge < 25 Then loading += _rates.YoungDriverLoading
        If input.Use = VehicleUse.Commercial Then
            loading += If(input.EngineCc > _rates.LargeEngineCc,
                          _rates.LargeEngineLoading, _rates.CommercialLoading)
        End If

        Dim discount = NoClaimDiscountFor(input)
        Dim net = RoundBaht(basePremium * (1D + loading) * (1D - discount))
        Dim duty = RoundBaht(net * _rates.StampDutyRate)
        Dim vat = RoundBaht((net + duty) * _rates.VatRate)
        Return New PremiumBreakdown(RoundBaht(basePremium), loading, discount, net, duty, vat)
    End Function

    ' The no-claim ladder rewards claim-free licence years; its last rung covers 5 years and more.
    Private Function NoClaimDiscountFor(input As RatingInput) As Decimal
        If input.ClaimsLast5Years > 0 Then Return 0D
        Return _rates.NoClaimDiscounts(
            Math.Min(input.LicenceYears, _rates.NoClaimDiscounts.Length - 1))
    End Function

    ' Math.Round defaults to banker's rounding (half to even): money needs AwayFromZero.
    Private Shared Function RoundBaht(amount As Decimal) As Decimal
        Return Math.Round(amount, 2, MidpointRounding.AwayFromZero)
    End Function
#End Region

    Public Function BaseRateFor(coverage As CoverageClass) As Decimal
        Select Case coverage
            Case CoverageClass.Class1 : Return _rates.Class1Rate
            Case CoverageClass.Class2Plus : Return _rates.Class2PlusRate
            Case CoverageClass.Class3Plus : Return _rates.Class3PlusRate
            Case Else : Return _rates.Class3Rate
        End Select
    End Function
End Class

File: lesson-10-testing-tdd-coverage/samples/L10.Pricing/PremiumCalculator.cs
namespace L10.Pricing;

/// <summary>Rating rules for a motor quote — the track's canonical tariff (see the curriculum).
/// Illustrative rates, not a real tariff.</summary>
public sealed class PremiumCalculator
{
    #region base-rate
    public static decimal BaseRate(CoverageClass coverage) => coverage switch
    {
        CoverageClass.Class1 => 0.021m,
        CoverageClass.Class2Plus => 0.012m,
        CoverageClass.Class3Plus => 0.009m,
        CoverageClass.Class3 => 0.004m,
        // enums are open: (CoverageClass)99 compiles, so this arm is reachable
        _ => throw new ArgumentOutOfRangeException(nameof(coverage), coverage, "unknown class"),
    };
    #endregion

    #region claims-guard
    public static decimal ClaimsLoading(int claims) =>
        claims switch
        {
            0 => 0.00m,
            1 => 0.10m,
            2 => 0.25m,
            // 3 or more in five years is not a price, it is a refusal
            _ => throw new QuoteDeclinedException(
                "3+ claims in 5 years"),
        };
    #endregion

    #region calculate
    public Premium Calculate(QuoteRequest request)
    {
        ArgumentNullException.ThrowIfNull(request);
        var (vehicle, driver) = (request.Vehicle, request.Driver);

        var loading = 1.00m + ClaimsLoading(driver.ClaimsLast5Years);
        if (AgeAt(driver.DateOfBirth, request.StartDate) < 25)
            loading += 0.20m;                                   // young driver
        if (vehicle.Use == VehicleUse.Commercial)
            loading += vehicle.EngineCc > 3_000 ? 0.35m : 0.25m; // commercial use

        var claimFreeYears = driver.ClaimsLast5Years == 0 ? driver.LicenceYears : 0;
        var gross = vehicle.SumInsured.Amount * BaseRate(request.Coverage) * loading;
        var net = Round(gross * (1 - NoClaimBonus.DiscountFor(claimFreeYears)));

        var duty = Round(net * 0.004m);                         // stamp duty 0.4%, to the satang
        var vat = Round((net + duty) * 0.07m);                  // VAT 7% on net + duty
        return new Premium(Thb(net), Thb(duty), Thb(vat), Thb(net + duty + vat));
    }
    #endregion

    #region age
    public static int AgeAt(DateOnly dateOfBirth, DateOnly on)
    {
        var age = on.Year - dateOfBirth.Year;
        return dateOfBirth > on.AddYears(-age) ? age - 1 : age;
    }
    #endregion

    private static decimal Round(decimal value) =>
        Math.Round(value, 2, MidpointRounding.AwayFromZero);

    private static Money Thb(decimal amount) => new(amount, "THB");
}
